File: pkg/adapter/http/middleware.go
package http

import (
	"log/slog"
	"net/http"
	"runtime/debug"
	"time"
)

// Recover guards downstream handlers from panics. A recovered panic is logged
// and translated to a 500 response with an INVALID_REQUEST-shaped JSON body so
// clients see consistent error envelopes.
func Recover(log *slog.Logger) Middleware {
	if log == nil {
		log = slog.Default()
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			defer func() {
				if rec := recover(); rec != nil {
					log.Error("http: panic recovered",
						"path", r.URL.Path,
						"method", r.Method,
						"panic", rec,
						"stack", string(debug.Stack()),
					)
					writeErrorJSON(w, http.StatusInternalServerError, "INTERNAL_ERROR", "internal server error")
				}
			}()
			next.ServeHTTP(w, r)
		})
	}
}

// Logging emits a structured log line per request after the handler returns.
// Status code is captured via a thin ResponseWriter wrapper.
func Logging(log *slog.Logger) Middleware {
	if log == nil {
		log = slog.Default()
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()
			lw := &loggingWriter{ResponseWriter: w, status: http.StatusOK}
			next.ServeHTTP(lw, r)
			log.Info("http request",
				"method", r.Method,
				"path", r.URL.Path,
				"status", lw.status,
				"duration_ms", time.Since(start).Milliseconds(),
			)
		})
	}
}

// loggingWriter captures the status code so Logging can report it. It also
// implements http.Flusher passthrough so SSE handlers downstream still flush.
type loggingWriter struct {
	http.ResponseWriter
	status      int
	wroteHeader bool
}

func (w *loggingWriter) WriteHeader(code int) {
	if !w.wroteHeader {
		w.status = code
		w.wroteHeader = true
	}
	w.ResponseWriter.WriteHeader(code)
}

func (w *loggingWriter) Write(b []byte) (int, error) {
	if !w.wroteHeader {
		w.wroteHeader = true
	}
	return w.ResponseWriter.Write(b)
}

// Flush forwards to the underlying writer when supported. Required so
// http.NewResponseController(w).Flush() in stream handlers still works when
// Logging is in the chain.
func (w *loggingWriter) Flush() {
	if f, ok := w.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

// Unwrap exposes the underlying ResponseWriter so http.NewResponseController
// can find the real Flusher / Hijacker / etc.
func (w *loggingWriter) Unwrap() http.ResponseWriter { return w.ResponseWriter }

File: sdk/python/src/openagentio/adapter/http/middleware.py
"""Raw ASGI middleware mirroring pkg/adapter/http/middleware.go.

Uses raw ASGI rather than Starlette's BaseHTTPMiddleware to avoid buffering
streaming (SSE) responses.
"""
from __future__ import annotations

import logging
import time
import traceback
from typing import Any, Callable, Awaitable

from starlette.types import ASGIApp, Receive, Scope, Send

from openagentio.adapter.http.errors import write_error_json
from openagentio.event.payload import CodeCodecFailure

ASGIMiddleware = Callable[[ASGIApp], ASGIApp]


class Recover:
    """Guards downstream from unhandled exceptions.

    A recovered exception is logged and translated to a 500 JSON response.
    """

    def __init__(self, logger: logging.Logger | None = None) -> None:
        self._log = logger or logging.getLogger("openagentio.http")

    def __call__(self, app: ASGIApp) -> ASGIApp:
        log = self._log

        async def middleware(scope: Scope, receive: Receive, send: Send) -> None:
            if scope["type"] not in ("http",):
                await app(scope, receive, send)
                return

            error_sent = False

            async def send_wrapper(message: dict[str, Any]) -> None:
                nonlocal error_sent
                if message.get("type") == "http.response.start":
                    error_sent = False
                await send(message)

            try:
                await app(scope, receive, send_wrapper)
            except Exception as exc:
                if error_sent:
                    return
                log.error(
                    "http: panic recovered",
                    extra={"path": scope.get("path", ""), "exc": exc},
                    exc_info=True,
                )
                resp = write_error_json(500, "INTERNAL_ERROR", "internal server error")
                error_sent = True
                await resp(scope, receive, send)

        return middleware


class Logging:
    """Emits a structured log line per request after the handler returns.

    Captures the status code via a thin ``send`` wrapper.
    """

    def __init__(self, logger: logging.Logger | None = None) -> None:
        self._log = logger or logging.getLogger("openagentio.http")

    def __call__(self, app: ASGIApp) -> ASGIApp:
        log = self._log

        async def middleware(scope: Scope, receive: Receive, send: Send) -> None:
            if scope["type"] not in ("http",):
                await app(scope, receive, send)
                return

            start = time.monotonic()
            status = 200

            async def send_wrapper(message: dict[str, Any]) -> None:
                nonlocal status
                if message.get("type") == "http.response.start":
                    raw_status = message.get("status", 200)
                    status = raw_status
                await send(message)

            await app(scope, receive, send_wrapper)
            duration_ms = (time.monotonic() - start) * 1000
            log.info(
                "http request",
                extra={
                    "method": scope.get("method", ""),
                    "path": scope.get("path", ""),
                    "status": status,
                    "duration_ms": round(duration_ms),
                },
            )

        return middleware
